CLOUDP-59033: MongoDB URI Added To Status

chatton opened the pull request:
### All Submissions:

* [ ] Have you opened an Issue before filing this PR?
* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).


This PR is based on top of #26 
This PR adds a Status Subresource to the MongoDB Resource.

There were issues with concurrent modification failures in the tests so I've added some retry logic to account for this.

The E2E tests were updated to test the status was updated correctly

EVG run: https://evergreen.mongodb.com/version/5e7a091ed1fe077fc3e379f1

chatton commented:
https://evergreen.mongodb.com/version/5e7a091ed1fe077fc3e379f1

chatton commented:
@LouisPlisso  can you take a quick look when you get a chance?

rodrigovalin commented on pkg/apis/mongodb/v1/mongodb_types.go:
```diff
@@ -31,6 +37,9 @@ type MongoDBSpec struct {
 // MongoDBStatus defines the observed state of MongoDB
 type MongoDBStatus struct {
+	MongoURI string `json:"mongoUri"`
+	Members  int    `json:"members"`
```
If this field is going to be a copy of what we have in `spec`, then it should be removed.

rodrigovalin commented on pkg/controller/mongodb/mongodb_controller.go:
```diff
@@ -124,19 +126,52 @@ func (r *ReplicaSetReconciler) Reconcile(request reconcile.Request) (reconcile.R
+		log.Infof("StatefulSet successfully Created/Updated")
+	}
+
+	log.Infof("waiting for StatefulSet %s/%s to reach ready state", mdb.Namespace, mdb.Name)
```
Debug message maybe?

rodrigovalin commented on pkg/controller/mongodb/mongodb_controller.go:
```diff
@@ -124,19 +126,52 @@ func (r *ReplicaSetReconciler) Reconcile(request reconcile.Request) (reconcile.R
+		return statefulset.IsReady(set)
+	})
+
+	if timedOut {
```
This kind of information is only relevant in a debugging scenario. Moreover, I think we can go with a different strategy here. We should rely on reconcile loop for the mechanics of object polling & waiting, not on our own loops with sleeps in between.

chatton replied:
The idea here was a relatively short wait period before retrying. We could of course shorten it to almost no wait period, this would end up with a lot of reconciliations consistently requeued (maybe that's not a problem though!)

rodrigovalin replied:
In the Kubernetes environment, a blocking controller is a problem.

rodrigovalin commented on pkg/controller/mongodb/mongodb_controller.go:
```diff
@@ -124,19 +126,52 @@ func (r *ReplicaSetReconciler) Reconcile(request reconcile.Request) (reconcile.R
 
+func (r ReplicaSetReconciler) updateStatusSuccess(mdb *mdbv1.MongoDB) error {
+	mdb.UpdateSuccess()
+	if err := r.client.Status().Update(context.TODO(), mdb); err != nil {
```
Will this mean that the object will go to error state if the Status object was not updated?

chatton replied:
No, at the moment there's no error state for the resource

rodrigovalin commented on test/e2e/e2eutil.go:
```diff
@@ -29,15 +29,28 @@ func RegisterTypesWithFramework(newTypes ...runtime.Object) error {
-			return f.Global.Client.Create(context.TODO(), mdb, &f.CleanupOptions{TestContext: ctx})
+// UpdateMongoDBResource applies the provided function to the most recent version of the MongoDB resource
+// and retries when there are conflicts
+func UpdateMongoDBResource(original *mdbv1.MongoDB, updateFunc func(*mdbv1.MongoDB)) error {
```
Can you add some sort of exponential backoff here?

chatton replied:
I don't think we need the backoff here, this loop is solely to account for period occurrences of apiserver errors that we get when the underlying resources are being modified elsewhere.

rodrigovalin replied:
In a scenario where apiserver errors are occurring, the less we want is to overwhelm the apiserver with a burst of requests.

rodrigovalin replied:
I think the loop is not needed here. Why would the object be modified by multiple actors?

rodrigovalin replied:
discussed:

- understand why there are multiple concurrent modifications, and fix that. but this 0..3 tries is tech debt.

chatton replied:
replaced with a single attempt, will keep track of any errors from this

rodrigovalin commented on test/e2e/e2eutil.go:
```diff
@@ -29,15 +29,28 @@ func RegisterTypesWithFramework(newTypes ...runtime.Object) error {
+			return err
+		}
+
+		updateFunc(original)
```
The original object is passed to the function, then it is being fetched from the API, then it is modified in place, then it is updated...

Unfortunatelly this means that the same object could be updated multiple times, and the updated object will be fetched from the API...

The "original" `original` object will have to be saved so every iteration is independant from the next.

rodrigovalin replied:
resolved in person

chatton commented:
@rodrigovalin 

* Adjusted the get/update function to perform a single operation
*  No longer polling for stateful set (also removed polling function as it isn't used anywhere else)
